taichi/backends/cuda/cupti_toolkit_functions.h

```diff
@@ -0,0 +1,203 @@
+#pragma once
+
+#if defined(TI_WITH_CUDA_TOOLKIT)
+
+#include <cupti_target.h>
+#include <cupti_result.h>
+#include <cupti_profiler_target.h>
+#include <nvperf_host.h>
+#include <nvperf_cuda_host.h>
+#include <nvperf_target.h>
+
+template <typename T>
+
+class ScopeExit {
+ public:
+  ScopeExit(T t) : t(t) {
+  }
+  ~ScopeExit() {
+    t();
+  }
+  T t;
+};
+
+template <typename T>
+ScopeExit<T> MoveScopeExit(T t) {
+  return ScopeExit<T>(t);
+};
+
+#define CUPTI_API_CALL(apiFuncCall)                                        \
+  do {                                                                     \
+    CUptiResult _status = apiFuncCall;                                     \
+    if (_status != CUPTI_SUCCESS) {                                        \
+      const char *errstr;                                                  \
+      cuptiGetResultString(_status, &errstr);                              \
+      fprintf(stderr, "%s:%d: error: function %s failed with error %s.\n", \
+              __FILE__, __LINE__, #apiFuncCall, errstr);                   \
+      exit(-1);                                                            \
+    }                                                                      \
+  } while (0)
+
+#define NVPW_API_CALL(apiFuncCall)                                         \
+  do {                                                                     \
+    NVPA_Status _status = apiFuncCall;                                     \
+    if (_status != NVPA_STATUS_SUCCESS) {                                  \
+      fprintf(stderr, "%s:%d: error: function %s failed with error %d.\n", \
+              __FILE__, __LINE__, #apiFuncCall, _status);                  \
+      exit(-1);                                                            \
+    }                                                                      \
+  } while (0)
+
+#define NV_ANONYMOUS_VARIABLE_DIRECT(name, line) name##line
+
+#define NV_ANONYMOUS_VARIABLE_INDIRECT(name, line) \
+  NV_ANONYMOUS_VARIABLE_DIRECT(name, line)
+
+#define SCOPE_EXIT(func)                                      \
+  const auto NV_ANONYMOUS_VARIABLE_INDIRECT(EXIT, __LINE__) = \
+      MoveScopeExit([=]() { func; })
+
+#define RETURN_IF_NVPW_ERROR(retval, actual)                 \
+  do {                                                       \
+    NVPA_Status status = actual;                             \
+    if (NVPA_STATUS_SUCCESS != status) {                     \
+      fprintf(stderr, "FAILED: %s with error %s\n", #actual, \
+              GetNVPWResultString(status));                  \
+      return retval;                                         \
+    }                                                        \
+  } while (0)
+
+static const char *GetNVPWResultString(NVPA_Status status) {
+  const char *errorMsg = NULL;
+  switch (status) {
```

Comment:
Use a macro for this?

Reply:
```
static const char *get_nvpw_result_string(NVPA_Status status) {
  const char *error_msg = NULL;
  switch (status) {
    NVPW_STATUS_RESULT(error_msg, NVPA_STATUS_ERROR)
    NVPW_STATUS_RESULT(error_msg, NVPA_STATUS_INTERNAL_ERROR)
    NVPW_STATUS_RESULT(error_msg, NVPA_STATUS_NOT_INITIALIZED)
    ...
```

Reply:
do you still need to pass in error_msg?

Reply:
no longer needed in `get_nvpw_result_string()` :P 
fixed in a9de3a5 , thanks!

Reply:
Oh sorry another small nit: for macros that you don't want to export to other files (since this is a header it can be included in many files), we usually #undef the macro immediately after its use to avoid unexpected results. These macros are less risky since their names are pretty complicated, but it's usually good practice to think about whether you still want to use it and undef it once it's done. :D

Reply:
Many Thanks!

Reply:
Many Thanks!